File: LeetCode/src/main/java/com/jack/hashtable/LeetCode49.java
package com.jack.hashtable;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class LeetCode49 {
    public static void main(String[] args) {
        
    }

    static class Solution{
        public List<List<String>> groupAnagrams(String[] strs) {

            Map<String, List<String>> map = new HashMap<>();
            List<List<String>> res = new ArrayList<>();

            for(String str :strs){
                String sortStr = sortString(str);
                if(!map.containsKey(sortStr)){
                    map.put(sortStr, new ArrayList<String>());
                }
                map.get(sortStr).add(str);
            }

            for(List<String> sList: map.values()){
                res.add(sList);
            }

            return res;
            
        }

        private String sortString(String str){
            char [] sArr = str.toCharArray();
            Arrays.sort(sArr);
            return new String(sArr);
        }
    }

}

File: LeetCode/src/main/java/com/jack/hashtable/LeetCode49.py
from collections import defaultdict
from typing import List


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        d = defaultdict(list)
        for s in strs:
            k = "".join(sorted(s))
            d[k].append(s)
        return list(d.values())


if __name__ == "__main__":
    pass 